File: src/my_ude_py/scripts/navigation_client.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from rclpy.action import ActionClient
from functools import partial

from ude_interface.action import Navigate
from std_msgs.msg import Empty


class NavigationClientNode(Node): 
    def __init__(self):
        super().__init__("navigation_client") 
        self.subscriber_ = self.create_subscription(Empty, "cancel_navigate",self.callback_cancel_goal,1)
        self.call_navigate_server(2.0, 3.0, 0.0)
        self.get_logger().info("Navigation Client Node has been started")

    def callback_cancel_goal(self, msg):
        self.get_logger().info('Canceling goal')
        # Cancel the goal
        future = self._goal_handle.cancel_goal_async()
        future.add_done_callback(self.cancel_done)

    def cancel_done(self, future):
        cancel_response = future.result()
        if len(cancel_response.goals_canceling) > 0:
            self.get_logger().info('Goal successfully canceled')
        else:
            self.get_logger().info('Goal failed to cancel')

    def call_navigate_server(self, x, y, z):
        self.action_client_=ActionClient(self, Navigate, 'navigate')
        while not self.action_client_.wait_for_server(1.0):
            self.get_logger().warn("Waiting for server Action... ")

        goal_msg = Navigate.Goal()
        goal_msg.goal_point.x = float(x)
        goal_msg.goal_point.y = float(y)
        goal_msg.goal_point.z = float(z)

        self.get_logger().info("Sending goal")

        send_goal_future = self.action_client_.send_goal_async(goal_msg, feedback_callback=self.feedback_callback)
        send_goal_future.add_done_callback(partial(self.goal_response_callback, x=x, y=y, z=z))

    def goal_response_callback(self, future, x, y, z):
        goal_handle = future.result()
        status_=goal_handle.status
        if not goal_handle.accepted:
            self.get_logger().error("Goal was rejected by server")
            return
        self.get_logger().info("Goal accepted by server, waiting for result...")                    
        self.get_logger().info("Status is: " + str(status_) + ", X is : " + str(x) +
                               ", Y is : " + str(y) + ", Z is : " + str(z))
        self._goal_handle = goal_handle

        self._get_result_future = goal_handle.get_result_async()
        self._get_result_future.add_done_callback(partial(self.get_result_callback, x=x, y=y, z=z))

    def get_result_callback(self, future, x, y, z):
        try:
            status = future.result().status
            result = future.result().result
            if(status==0):
                self.get_logger().warn("Unknown result code")  
            elif(status==5):
                self.get_logger().warn("Goal was canceled")  
            elif(status==4):
                self.get_logger().info("Goal was succeeded")  
            elif(status==6):
                self.get_logger().warn("Goal was aborted")  

            self.get_logger().info("Status is: " + str(status) + ", X is : " + str(x) +
                               ", Y is : " + str(y) + ", Z is : " + str(z))
            self.get_logger().info('Received result:  {0}'.format(result.elapsed_time))

        except Exception as e:
            self.get_logger().error("Service call failed %r" % (e,))

    def feedback_callback(self, feedback_msg):
        feedback = feedback_msg.feedback
        self.get_logger().info('Received feedback: {0}'.format(feedback.distance_to_point))


def main(args=None):
    rclpy.init(args=args)
    node = NavigationClientNode() 
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: src/my_ude_cpp/src/navigation_client.cpp
#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"
#include "example_interfaces/msg/empty.hpp"

#include "ude_interface/action/navigate.hpp"
#include "geometry_msgs/msg/point.hpp"

using std::placeholders::_1;
using std::placeholders::_2;
using geometry_msgs::msg::Point;

typedef ude_interface::action::Navigate Navigate;
typedef rclcpp_action::ClientGoalHandle<Navigate> GoalHandle;
typedef rclcpp_action::Client<Navigate>::CancelResponse CancelResponse; 

class NavigationClientNode : public rclcpp::Node 
{
public:
    NavigationClientNode() : Node("navigation_client") 
    {
        subscriber_ = this-> create_subscription<example_interfaces::msg::Empty>("cancel_navigate",1,
                std::bind(&NavigationClientNode::callback_cancel_goal, this, std::placeholders::_1));
        action_client_ =  rclcpp_action::create_client<Navigate>(this, "navigate");
        call_navigate_server(3.0, 5.0, 0.0);
    	RCLCPP_INFO(this->get_logger(), "Navigation Client Node has been started.");
    }

private:
    void callback_cancel_goal(const example_interfaces::msg::Empty::SharedPtr msg)
    {
        (void)msg;
        RCLCPP_INFO(this->get_logger(), "Canceling goal");
        action_client_->async_cancel_goal(this->goal_handle_,std::bind(&NavigationClientNode::cancel_done, this, _1));
    }

    void cancel_done(const CancelResponse::SharedPtr & cancel_response){
        switch (cancel_response->return_code) {
        case CancelResponse::ERROR_NONE:
            RCLCPP_INFO(this->get_logger(), "Goal successfully canceled");
            break;
        case CancelResponse::ERROR_REJECTED:
            RCLCPP_WARN(this->get_logger(), "Goal reject to cancel");
            break;
        default:
            RCLCPP_WARN(this->get_logger(), "Goal failed to cancel");
            break;
        }
    }

    void call_navigate_server(float x, float y, float z)
    {
        
        while (!action_client_->wait_for_action_server(std::chrono::seconds(1))) {
            RCLCPP_WARN(this->get_logger(), "Action server not available after waiting");
        }
        auto goal_msg = Navigate::Goal();
        goal_msg.goal_point.x = x;
        goal_msg.goal_point.y = y;
        goal_msg.goal_point.z = z;

        RCLCPP_INFO(this->get_logger(), "Sending goal");
        auto send_goal_options = rclcpp_action::Client<Navigate>::SendGoalOptions();
        send_goal_options.goal_response_callback = std::bind(&NavigationClientNode::goal_response_callback, this, _1, 
            x, y, z);
        send_goal_options.feedback_callback = std::bind(&NavigationClientNode::feedback_callback, this, _1, _2);
        send_goal_options.result_callback = std::bind(&NavigationClientNode::result_callback, this, _1, x, y, z);
        action_client_->async_send_goal(goal_msg, send_goal_options);
    }

    void goal_response_callback(const GoalHandle::SharedPtr & goal_handle, float x, float y, float z)
    {
        if (!goal_handle) {
            RCLCPP_ERROR(this->get_logger(), "Goal was rejected by server");
        } else {
            auto status_= goal_handle->get_status();
            RCLCPP_INFO(this->get_logger(), "Goal accepted by server, waiting for result...");
            RCLCPP_INFO(this->get_logger(), "Status is: %d, X is : %f, Y is :  %f, Z is :  %f", status_ ,x, y, z);
            this->goal_handle_=goal_handle;
        }
    }

    void feedback_callback(GoalHandle::SharedPtr, const std::shared_ptr<const Navigate::Feedback> feedback)
    {
        float distance_to_point = feedback->distance_to_point;
        RCLCPP_INFO(this->get_logger(), "Received feedback: %f", distance_to_point);
    }

    void result_callback(const GoalHandle::WrappedResult & result, float x, float y, float z)
    {
        auto status_=result.code;
        switch (result.code) {
        case rclcpp_action::ResultCode::SUCCEEDED:
            RCLCPP_INFO(this->get_logger(), "Goal was succeeded");
            break;
        case rclcpp_action::ResultCode::ABORTED:
            RCLCPP_WARN(this->get_logger(), "Goal was aborted");
            break;
        case rclcpp_action::ResultCode::CANCELED:
            RCLCPP_WARN(this->get_logger(), "Goal was canceled");
            break;
        default:
            RCLCPP_WARN(this->get_logger(), "Unknown result code");  
            break;
        }
        RCLCPP_INFO(this->get_logger(), "Status is: %d, X is : %f, Y is :  %f, Z is :  %f", int(status_) ,x, y, z);
        int elapsed_time_f = result.result->elapsed_time;
        RCLCPP_INFO(this->get_logger(), "Received result: %d", elapsed_time_f);
    }


    rclcpp::Subscription<example_interfaces::msg::Empty>::SharedPtr subscriber_;
    GoalHandle::SharedPtr goal_handle_;
    rclcpp_action::Client<Navigate>::SharedPtr action_client_;
    std::vector<std::thread> threads_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<NavigationClientNode>(); 
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}